ext/nnstreamer/tensor_decoder/tensordec-python3.cc

```diff
@@ -0,0 +1,525 @@
+    ret = GST_FLOW_ERROR;
+  }
+
+done:
+  Py_UNLOCK ();
+  return ret;
+}
+
+/**
+ * @brief	get output caps
+ */
+GstCaps *
+PYDecoderCore::getOutCaps (const GstTensorsConfig *config)
+{
+  PyObject *result = NULL;
+  GstCaps *caps = NULL;
+
+  Py_LOCK ();
+  if (!PyObject_HasAttrString (core_obj, (char *)"getOutCaps")) {
+    ml_loge ("Cannot find 'getOutCaps'");
+    ml_loge ("defualt caps is `application/octet-stream`");
+    caps = gst_caps_from_string ("application/octet-stream");
+    goto done;
+  }
+
+  result = PyObject_CallMethod (core_obj, (char *)"getOutCaps", NULL);
+  if (result) {
+    gchar * caps_str = PyBytes_AsString (result);
+    caps = gst_caps_from_string (caps_str);
+    Py_XDECREF (result);
+  } else {
+    caps = gst_caps_from_string ("application/octet-stream");
+  }
+
+done:
+  Py_UNLOCK ();
+  return caps;
+}
+
+/**
+ * @brief	load the python script
+ */
+int
+PYDecoderCore::loadScript ()
+{
+  PyObject *module = PyImport_ImportModule (module_name.c_str ());
+
+  if (module) {
+    PyObject *cls = PyObject_GetAttrString (module, "CustomDecoder");
+    if (cls) {
+      core_obj = PyObject_CallObject (cls, NULL);
+      Py_XDECREF (cls);
+    } else {
+      Py_ERRMSG ("Cannot find 'CustomDecoder' class in the script\n");
+      return -2;
+    }
+    Py_XDECREF (module);
+  } else {
+    Py_ERRMSG ("the script is not properly loaded\n");
+    return -1;
+  }
+
+  return 0;
+}
+
+/**
+ * @brief	initialize the object with python script
+ */
+int
+PYDecoderCore::init ()
+{
+  /** Find nnstreamer_api module */
+  PyObject *api_module = PyImport_ImportModule ("nnstreamer_python");
+  if (api_module == NULL) {
+    return -EINVAL;
+  }
+
+  shape_cls = PyObject_GetAttrString (api_module, "TensorShape");
+  Py_XDECREF (api_module);
+
+  if (shape_cls == NULL)
+    return -EINVAL;
+
+  return loadScript ();
+}
+
+/**
+ * @brief	get the script path
+ * @return the script path.
+ */
+const char *
+PYDecoderCore::getScriptPath ()
+{
+  return script_path.c_str ();
+}
+
+/**
+ * @brief	allocate TensorShape object
+ * @param info : the tensor info
+ * @return created object
+ */
+PyObject *
+PYDecoderCore::PyTensorShape_New (const GstTensorInfo *info)
+{
+  PyObject *args = PyTuple_New (2);
+  PyObject *dims = PyList_New (0);
+  PyObject *type = (PyObject *)PyArray_DescrFromType (getNumpyType (info->type));
+
+  if (nullptr == args || nullptr == dims || nullptr == type)
+    throw std::runtime_error ("PYCore::PyTensorShape_New() has failed (1).");
+
+  for (int i = 0; i < NNS_TENSOR_RANK_LIMIT; i++)
+    PyList_Append (dims, PyLong_FromLong ((uint64_t)info->dimension[i]));
+
+  PyTuple_SetItem (args, 0, dims);
+  PyTuple_SetItem (args, 1, type);
+
+  return PyObject_CallObject (shape_cls, args);
+  /* Its value is checked by setInputTensorDim */
+}
+
+/**
+ * @brief	return the data type of the tensor for Python numpy
+ * @param tType	: the defined type of NNStreamer
+ * @return the enum of defined numpy datatypes
+ */
+NPY_TYPES
```

Comment:
Can you make this routine common across decoder and converter?

Reply:
And also in filter?